core/src/components/datetime/utils/manipulation.ts

```diff
@@ -424,23 +424,52 @@ export const validateParts = (
  * Returns the closest date to refParts
  * that also meets the constraints of
  * the *Values params.
- * @param refParts The reference date
- * @param monthValues The allowed month values
- * @param dayValues The allowed day (of the month) values
- * @param yearValues The allowed year values
- * @param hourValues The allowed hour values
- * @param minuteValues The allowed minute values
  */
-export const getClosestValidDate = (
-  refParts: DatetimeParts,
-  monthValues?: number[],
-  dayValues?: number[],
-  yearValues?: number[],
-  hourValues?: number[],
-  minuteValues?: number[]
-) => {
+export const getClosestValidDate = ({
+  refParts,
+  monthValues,
+  dayValues,
+  yearValues,
+  hourValues,
+  minuteValues,
+  minParts,
+  maxParts,
+}: {
+  /**
+   * The reference date
+   */
+  refParts: DatetimeParts;
+  /**
+   * The allowed month values
+   */
+  monthValues?: number[];
+  /**
+   * The allowed day (of the month) values
+   */
+  dayValues?: number[];
+  /**
+   * The allowed year values
+   */
+  yearValues?: number[];
+  /**
+   * The allowed hour values
+   */
+  hourValues?: number[];
+  /**
+   * The allowed minute values
+   */
+  minuteValues?: number[];
+  /**
+   * The minimum date that can be returned
+   */
+  minParts?: DatetimeParts;
+  /**
+   * The maximum date that can be returned
+   */
+  maxParts?: DatetimeParts;
+}) => {
   const { hour, minute, day, month, year } = refParts;
-  const copyParts = { ...refParts, dayOfWeek: undefined };
+  const copyParts = clampDate({ ...refParts, dayOfWeek: undefined }, minParts, maxParts);
```

Comment:
This implementation does not appear to be correct. `clampDate` will ensure that `copyParts` is within the min/max parts, but then the `findClosestValue` below will potentially move it outside the min/max bounds.

Example:

```js
getClosestValidDate({
    refParts: { month: 10, day: 10, year: 2023 },
    monthValues: [2, 3, 7, 9],
    dayValues: [4, 15, 25],
    yearValues: [2020, 2021, 2023],
    maxParts: {
      year: 2021,
      month: 9,
      day: 14
    },
    minParts: {
      year: 2021,
      month: 4,
      day: 10
    }
  })
```

This will return "2023-09-15" even though the max is "2021-09-14".

Reply:
I have a somewhat gnarly implementation that works with test cases. This code can likely be simplified.

```js
const { hour, minute, day, month, year } = refParts;
  const copyParts = { ...refParts, dayOfWeek: undefined };

  if (yearValues !== undefined) {
    const filteredYearValues = yearValues.filter((year) => {
      /**
       * If the allowed year value is outside of the min
       * year then it should not be considered.
       */
      if (minParts !== undefined && year < minParts.year) {
        return false;
      }

      /**
       * If the allowed year value is outside of the max
       * year then it should not be considered.
       */
      if (maxParts !== undefined && year > maxParts.year) {
        return false;
      }

      return true;
    });
    copyParts.year = findClosestValue(year, filteredYearValues);
  }

  if (monthValues !== undefined) {
    const filteredMonthValues = monthValues.filter((month) => {
      /**
       * If the allowed month value is outside of the min
       * month then it should not be considered only if the
       * chosen year matches the min year.
       */
      if (minParts !== undefined && copyParts.year === minParts.year && month < minParts.month) {
        return false;
      }

      /**
       * If the allowed month value is outside of the max
       * month then it should not be considered only if the
       * chosen year matches the max year.
       */
      if (maxParts !== undefined && copyParts.year === maxParts.year && month > maxParts.month) {
        return false;
      }

      return true;
    });
    copyParts.month = findClosestValue(month, filteredMonthValues);
  }

  // Day is nullable but cannot be undefined
  if (day !== null && dayValues !== undefined) {
    const filteredDayValues = dayValues.filter((day) => {
      /**
       * If the allowed day value is outside of the min
       * day then it should not be considered only if the
       * chosen year and month matches the min year and month.
       */
      if (
        minParts !== undefined &&
        copyParts.year === minParts.year &&
        copyParts.month === minParts.month &&
        minParts.day !== null &&
        day < minParts.day
      ) {
        return false;
      }

      /**
       * If the allowed day value is outside of the max
       * day then it should not be considered only if the
       * chosen year and month matches the max year and month.
       */
      if (
        maxParts !== undefined &&
        copyParts.year === maxParts.year &&
        copyParts.month === maxParts.month &&
        maxParts.day !== null &&
        day > maxParts.day
      ) {
        return false;
      }

      return true;
    });
    copyParts.day = findClosestValue(day, filteredDayValues);
  }

  if (hour !== undefined && hourValues !== undefined) {
    const filteredHourValues = hourValues.filter((hour) => {
      /**
       * If the allowed hour value is outside of the min
       * hour then it should not be considered only if the
       * chosen day is the same as the min day.
       */
      if (minParts?.hour !== undefined && isSameDay(copyParts, minParts) && hour < minParts.hour) {
        return false;
      }

      /**
       * If the allowed hour value is outside of the max
       * hour then it should not be considered only if the
       * chosen day is the same as the max day.
       */
      if (maxParts?.hour !== undefined && isSameDay(copyParts, maxParts) && hour > maxParts.hour) {
        return false;
      }

      return true;
    });
    copyParts.hour = findClosestValue(hour, filteredHourValues);
    copyParts.ampm = parseAmPm(copyParts.hour);
  }

  if (minute !== undefined && minuteValues !== undefined) {
    const filteredMinuteValues = minuteValues.filter((minute) => {
      /**
       * If the allowed minute value is outside of the min
       * minute then it should not be considered only if the
       * chosen day is the same as the min day and the chosen
       * hour is the same as the min hour.
       */
      if (
        minParts?.minute !== undefined &&
        isSameDay(copyParts, minParts) &&
        copyParts.hour === minParts.hour &&
        minute < minParts.minute
      ) {
        return false;
      }

      /**
       * If the allowed minute value is outside of the max
       * minute then it should not be considered only if the
       * chosen day is the same as the max day and the chosen
       * hour is the same as the max hour.
       */
      if (
        maxParts?.minute !== undefined &&
        isSameDay(copyParts, maxParts) &&
        copyParts.hour === maxParts.hour &&
        minute > maxParts.minute
      ) {
        return false;
      }

      return true;
    });
    copyParts.minute = findClosestValue(minute, filteredMinuteValues);
  }

  return copyParts;
```

Tests: 

```js
expect(
  getClosestValidDate({
    refParts: { month: 10, day: 10, year: 2023 },
    monthValues: [2, 3, 7, 9, 10],
    dayValues: [4, 15, 25],
    yearValues: [2020, 2021, 2023],
    maxParts: {
      year: 2021,
      month: 9,
      day: 14,
    },
    minParts: {
      year: 2021,
      month: 4,
      day: 10,
    },
  })
).toEqual({
  day: 4,
  month: 9,
  year: 2021,
  dayOfWeek: undefined,
});
expect(
  getClosestValidDate({
    refParts: { month: 4, day: 2, year: 2020, hour: 15, minute: 20 },
    monthValues: [4, 9, 11],
    dayValues: [11, 12, 13, 14],
    yearValues: [2020, 2021, 2023],
    hourValues: [9, 10, 11],
    minuteValues: [11, 12, 13, 14, 15],
    maxParts: {
      year: 2021,
      month: 9,
      day: 14,
      hour: 10,
      minute: 11,
    },
    minParts: {
      year: 2021,
      month: 9,
      day: 10,
      hour: 10,
      minute: 10,
    },
  })
).toEqual({
  day: 11,
  month: 9,
  year: 2021,
  hour: 11,
  minute: 15,
  ampm: 'am',
  dayOfWeek: undefined,
});
expect(
  getClosestValidDate({
    refParts: { month: 4, day: 2, year: 2020, hour: 15, minute: 20 },
    monthValues: [4, 9, 11],
    dayValues: [10, 12, 13, 14],
    yearValues: [2020, 2021, 2023],
    hourValues: [9, 10, 11],
    minuteValues: [11, 12, 13, 14, 15],
    maxParts: {
      year: 2021,
      month: 9,
      day: 10,
      hour: 10,
      minute: 15,
    },
    minParts: {
      year: 2021,
      month: 9,
      day: 10,
      hour: 10,
      minute: 10,
    },
  })
).toEqual({
  day: 10,
  month: 9,
  year: 2021,
  hour: 10,
  minute: 15,
  ampm: 'am',
  dayOfWeek: undefined,
});
```

Reply:
Pushed up a very close solution to this, just using the `isBefore` and `isAfter` utilities as well. 

Let me know if you think it would be helpful to include the comments here as well. With the use of the utilities & the tests I left them out for now.